### ISSUE TITLE
Extremely large (perplexity)loss when implementing memory network for language modeling 

### ISSUE THREAD

**Issue by nicklhy** [apache/incubator-mxnet#4377]
Recently, I am trying to implement the [End-To-End Memory Networks](http://arxiv.org/abs/1503.08895v4) for language modeling according to [carpedm20](https://github.com/carpedm20)'s Tensorflow [implementation](https://github.com/carpedm20/MemN2N-tensorflow) and you can find my code [here](https://github.com/nicklhy/MemN2N). But even after I set all the parameters same with his, my mxnet version doesn't work properly and produce extremely large (perplexity)loss in training phase from the very beginning.

I tried to change the learning rate and gradient clip to much smaller values(i.e. init_lr=0.0005, max_grad_norm=1). Now, the validation perplexity first fall to 200-300 but then increase to 400-500 again.

I also tried to use ADAM instead of SGD. This change could give a better validation result around 160 but still much higher than the TF implementation(114).

Currently, the only difference between these two implementations I know is the lr factor method. The TF version scales the learning rate when the loss stops to decrease. But I don't know how to implement it in MXNet. Instead, I use a normal mx.lr_scheduler.FactorScheduler.

Questions:
1. Is there any difference between tensorflow's SGD and mxnet's SGD(when both momentum and wd are set to 0)?
2. Why does the ADAM produces a not too bad result while the SGD fails to converge from the very beginning even their initial learning rates are the same.

**Comment by piiswrong** [comment]
data = mx.sym.Variable('data', shape=(-1, edim))
I don't think -1 works

**Comment by piiswrong** [comment]
try plotting your network definition and check against original implementation

**Comment by nicklhy** [comment]
@piiswrong , I didn't get any error message by setting shape=(-1, xxx) so far. Do you mean I should change it to batch size or just define these input variables without shape like `data = mx.sym.Variable('data')`?

I plot the network structure(when nhop=1) and you can find it [here](https://github.com/nicklhy/MemN2N/blob/master/cache/network_structure.png).

I know it is not a good idea to post such long code here and ask other people to help me debug. But I have been stuck here for a long time without any progress. Really need some help about the above mentioned problems.

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
